packages/repo-tools/src/commands/generate-patch/generate-patch.ts

```diff
@@ -0,0 +1,421 @@
+    {
+      cwd: ctx.sourceRepo.root.dir,
+    },
+  );
+  const describe = describeResult.stdout.toString('utf8').trim();
+  const name = `${cleanPackageName}-${version}-${describe}.patch`;
+  const patchDir = joinPath(ctx.targetRepo.root.dir, '.yarn', 'patches');
+
+  await fs.ensureDir(patchDir);
+  await fs.writeFile(joinPath(patchDir, name), patch, 'utf8');
+
+  const locator = `${ctx.sourcePkg.packageJson.name}@npm:${version}`;
+  return `patch:${encodeURIComponent(locator)}#${posix.join(
+    '.',
+    '.yarn',
+    'patches',
+    name,
+  )}`;
+}
+
+// Check if an existing resolution entry is a patch, and in that case return the
+// path to the patch file.
+function tryParsePatchResolution(value?: string): string | undefined {
+  if (!value) {
+    return undefined;
+  }
+  if (!value.startsWith('patch:')) {
+    return undefined;
+  }
+  const patchFilePath = value.split('#')[1];
+  return patchFilePath;
+}
+
+// Read out the descriptors for all entries of the package in the target repo
+async function readResolutionMap(
+  ctx: PatchContext,
+): Promise<Map<string, string>> {
+  const { stdout: whyOutput } = await exec(
+    'yarn',
+    ['why', '--json', ctx.packageName],
+    {
+      cwd: ctx.targetRoot,
+      maxBuffer: 1024 * 1024 * 1024,
+    },
+  );
+
+  const resolutionMap = new Map<string, string>();
+
+  for (const line of whyOutput.toString('utf8').trim().split(/\r?\n/)) {
+    for (const { locator, descriptor } of Object.values(
+      JSON.parse(line).children,
+    ) as Array<{ locator: string; descriptor: string }>) {
+      const existing = resolutionMap.get(descriptor);
+      if (existing) {
+        if (existing !== locator) {
+          throw new Error(
+            `Conflicting resolutions in target package for ${descriptor}: ${existing} vs ${locator}`,
+          );
+        }
+      } else {
+        resolutionMap.set(descriptor, locator);
+      }
+    }
+  }
+
+  return resolutionMap;
+}
+
+// Build and pack the source package to an archive in the work directory
+async function buildTargetArchive(ctx: PatchContext): Promise<string> {
+  await exec('yarn', ['build'], {
+    cwd: ctx.sourcePkg.dir,
+  });
+
+  const archiveName = 'target.tgz';
+
+  await exec('yarn', ['pack', '--out', joinPath(ctx.workDir, archiveName)], {
+    cwd: ctx.sourcePkg.dir,
+  });
+
+  return archiveName;
+}
+
+// Downloads a tar archive for the source package at the specified version
+async function downloadArchive(
+  ctx: PatchContext,
+  version: string,
+): Promise<string> {
+  const nameWithoutScope = ctx.packageName.replace(/^@.+\//, '');
+  const tarName = `${nameWithoutScope}-${version}.tgz`;
+
+  if (await fs.pathExists(joinPath(ctx.workDir, tarName))) {
+    return tarName;
+  }
+
+  try {
+    const res = await fetch(
+      `${ctx.registryUrl}/${ctx.packageName}/-/${tarName}`,
+    );
+    if (!res.ok) {
+      throw new Error(
+        `Request failed with status ${res.status} ${res.statusText}`,
```

Comment:
hm since it's such a complex generated url, maybe worth including it in the log line (i see package@version below as well, maybe it suffices, maybe not)